mrbgems/mruby-objectspace/src/mruby_objectspace.c

```diff
@@ -168,12 +176,206 @@ os_each_object(mrb_state *mrb, mrb_value self)
   return mrb_fixnum_value(d.count);
 }
 
+static void os_memsize_of_object(mrb_state*,mrb_value,mrb_bool,mrb_int*);
+
+static int
+os_memsize_ivar_cb(mrb_state *mrb, mrb_sym _name, mrb_value obj, void *data)
+{
+  mrb_int *cb_data = (mrb_int *)data;
+  mrb_int recurse = *(&cb_data[0]);
+  mrb_int* total = &cb_data[1];
+
+  os_memsize_of_object(mrb, obj, (mrb_bool)(recurse), total);
+  return 0;
+}
+
+static void
+os_memsize_of_ivars(mrb_state* mrb, mrb_value obj, mrb_bool recurse, mrb_int *t)
+{
+  /* need iv segment table size */
```

Comment:
Currently hash tables and their data segments are within `.c` files not accessible outside that translation unit. Is it okay to pull the definitions of iv and hash tables into header files? Otherwise I have to copy the definition of these structs into this file so we can calculate them.

Something like this

```patch
diff --git a/include/mruby/hash.h b/include/mruby/hash.h
index 0052a110..18dadab5 100644
--- a/include/mruby/hash.h
+++ b/include/mruby/hash.h
@@ -20,6 +20,33 @@ struct RHash {
   struct htable *ht;
 };
 
+/* hash table structure */
+
+struct h_segkv {
+  mrb_value key;
+  mrb_value val;
+};
+
+typedef struct h_segindex {
+  size_t size;
+  size_t capa;
+  struct h_segkv *table[];
+} h_segindex;
+
+typedef struct h_segment {
+  uint16_t size;
+  struct h_segment *next;
+  struct h_segkv e[];
+} h_segment;
+
+typedef struct htable {
+  h_segment *rootseg;
+  h_segment *lastseg;
+  mrb_int size;
+  uint16_t last_len;
+  h_segindex *index;
+} htable;
+
 #define mrb_hash_ptr(v)    ((struct RHash*)(mrb_ptr(v)))
 #define mrb_hash_value(p)  mrb_obj_value((void*)(p))
 
diff --git a/include/mruby/variable.h b/include/mruby/variable.h
index 6e918cf5..d671d4d9 100644
--- a/include/mruby/variable.h
+++ b/include/mruby/variable.h
@@ -29,6 +29,23 @@ struct global_entry {
   mrb_sym id;
 };
 
+#ifndef MRB_IV_SEGMENT_SIZE
+#define MRB_IV_SEGMENT_SIZE 4
+#endif
+
+typedef struct iv_segment {
+  mrb_sym key[MRB_IV_SEGMENT_SIZE];
+  mrb_value val[MRB_IV_SEGMENT_SIZE];
+  struct iv_segment *next;
+} iv_segment;
+
+/* Instance variable table structure */
+typedef struct iv_tbl {
+  iv_segment *rootseg;
+  size_t size;
+  size_t last_len;
+} iv_tbl;
+
 mrb_value mrb_vm_special_get(mrb_state*, mrb_sym);
 void mrb_vm_special_set(mrb_state*, mrb_sym, mrb_value);
 mrb_value mrb_vm_cv_get(mrb_state*, mrb_sym);
diff --git a/src/hash.c b/src/hash.c
index 4d531090..99dc66bf 100644
--- a/src/hash.c
+++ b/src/hash.c
@@ -21,39 +21,13 @@ mrb_int mrb_float_id(mrb_float f);
 #endif
 #define HT_SEG_INCREASE_RATIO 6 / 5
 
-struct segkv {
-  mrb_value key;
-  mrb_value val;
-};
-
-typedef struct segment {
-  uint16_t size;
-  struct segment *next;
-  struct segkv e[];
-} segment;
-
-typedef struct segindex {
-  size_t size;
-  size_t capa;
-  struct segkv *table[];
-} segindex;
-
-/* hash table structure */
-typedef struct htable {
-  segment *rootseg;
-  segment *lastseg;
-  mrb_int size;
-  uint16_t last_len;
-  segindex *index;
-} htable;
-
 static /* inline */ size_t
 ht_hash_func(mrb_state *mrb, htable *t, mrb_value key)
 {
   enum mrb_vtype tt = mrb_type(key);
   mrb_value hv;
   size_t h;
-  segindex *index = t->index;
+  h_segindex *index = t->index;
   size_t capa = index ? index->capa : 0;
 
   switch (tt) {
@@ -121,7 +95,7 @@ ht_hash_equal(mrb_state *mrb, htable *t, mrb_value a, mrb_value b)
 
   default:
     {
-      segindex *index = t->index;
+      h_segindex *index = t->index;
       size_t capa = index ? index->capa : 0;
       mrb_bool eql = mrb_eql(mrb, a, b);
       if (index && (index != t->index || capa != index->capa)) {
@@ -170,8 +144,8 @@ ht_index(mrb_state *mrb, htable *t)
 {
   size_t size = (size_t)t->size;
   size_t mask;
-  segindex *index = t->index;
-  segment *seg;
+  h_segindex *index = t->index;
+  h_segment *seg;
   size_t i;
 
   /* allocate index table */
@@ -180,7 +154,7 @@ ht_index(mrb_state *mrb, htable *t)
   }
   power2(size);
   if (!index || index->capa < size) {
-    index = (segindex*)mrb_realloc_simple(mrb, index, sizeof(segindex)+sizeof(struct segkv*)*size);
+    index = (h_segindex*)mrb_realloc_simple(mrb, index, sizeof(h_segindex)+sizeof(struct h_segkv*)*size);
     if (index == NULL) {
       mrb_free(mrb, t->index);
       t->index = NULL;
@@ -221,9 +195,9 @@ ht_index(mrb_state *mrb, htable *t)
 static void
 ht_compact(mrb_state *mrb, htable *t)
 {
-  segment *seg;
+  h_segment *seg;
   uint16_t i, i2;
-  segment *seg2 = NULL;
+  h_segment *seg2 = NULL;
   mrb_int size = 0;
 
   if (t == NULL) return;
@@ -277,8 +251,8 @@ ht_compact(mrb_state *mrb, htable *t)
   }
 }
 
-static segment*
-segment_alloc(mrb_state *mrb, segment *seg)
+static h_segment*
+h_segment_alloc(mrb_state *mrb, h_segment *seg)
 {
   uint32_t size;
 
@@ -288,7 +262,7 @@ segment_alloc(mrb_state *mrb, segment *seg)
     if (size > UINT16_MAX) size = UINT16_MAX;
   }
 
-  seg = (segment*)mrb_malloc(mrb, sizeof(segment)+sizeof(struct segkv)*size);
+  seg = (h_segment*)mrb_malloc(mrb, sizeof(h_segment)+sizeof(struct h_segkv)*size);
   seg->size = size;
   seg->next = NULL;
 
@@ -299,9 +273,9 @@ segment_alloc(mrb_state *mrb, segment *seg)
 static void
 ht_index_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
 {
-  segindex *index = t->index;
+  h_segindex *index = t->index;
   size_t k, sp, step = 0, mask;
-  segment *seg;
+  h_segment *seg;
 
   if (index->size >= UPPER_BOUND(index->capa)) {
     /* need to expand table */
@@ -331,8 +305,8 @@ ht_index_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
   if (t->last_len < seg->size) {
     index->table[k] = &seg->e[t->last_len++];
   }
-  else {                        /* append a new segment */
-    seg->next = segment_alloc(mrb, seg);
+  else {                        /* append a new h_segment */
+    seg->next = h_segment_alloc(mrb, seg);
     seg = seg->next;
     seg->next = NULL;
     t->lastseg = seg;
@@ -349,7 +323,7 @@ ht_index_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
 static void
 ht_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
 {
-  segment *seg;
+  h_segment *seg;
   mrb_int i, deleted = 0;
 
   if (t == NULL) return;
@@ -361,7 +335,7 @@ ht_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
   while (seg) {
     for (i=0; i<seg->size; i++) {
       mrb_value k = seg->e[i].key;
-      /* Found room in last segment after last_len */
+      /* Found room in last h_segment after last_len */
       if (!seg->next && i >= t->last_len) {
         seg->e[i].key = key;
         seg->e[i].val = val;
@@ -382,7 +356,7 @@ ht_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
   }
   /* Not found */
 
-  /* Compact if last segment has room */
+  /* Compact if last h_segment has room */
   if (deleted > 0 && deleted > MRB_HT_INIT_SIZE) {
     ht_compact(mrb, t);
   }
@@ -394,8 +368,8 @@ ht_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
     i = t->last_len;
   }
   else {
-    /* append new segment */
-    seg = segment_alloc(mrb, t->lastseg);
+    /* append new h_segment */
+    seg = h_segment_alloc(mrb, t->lastseg);
     i = 0;
     if (t->rootseg == NULL) {
       t->rootseg = seg;
@@ -417,7 +391,7 @@ ht_put(mrb_state *mrb, htable *t, mrb_value key, mrb_value val)
 static mrb_bool
 ht_index_get(mrb_state *mrb, htable *t, mrb_value key, mrb_value *vp)
 {
-  segindex *index = t->index;
+  h_segindex *index = t->index;
   size_t mask = HT_MASK(index);
   size_t k = ht_hash_func(mrb, t, key) & mask;
   size_t step = 0;
@@ -437,7 +411,7 @@ ht_index_get(mrb_state *mrb, htable *t, mrb_value key, mrb_value *vp)
 static mrb_bool
 ht_get(mrb_state *mrb, htable *t, mrb_value key, mrb_value *vp)
 {
-  segment *seg;
+  h_segment *seg;
   mrb_int i;
 
   if (t == NULL) return FALSE;
@@ -469,7 +443,7 @@ ht_get(mrb_state *mrb, htable *t, mrb_value key, mrb_value *vp)
 static mrb_bool
 ht_del(mrb_state *mrb, htable *t, mrb_value key, mrb_value *vp)
 {
-  segment *seg;
+  h_segment *seg;
   mrb_int i;
 
   if (t == NULL) return FALSE;
@@ -499,14 +473,14 @@ ht_del(mrb_state *mrb, htable *t, mrb_value key, mrb_value *vp)
 static void
 ht_foreach(mrb_state *mrb, htable *t, mrb_hash_foreach_func *func, void *p)
 {
-  segment *seg;
+  h_segment *seg;
   mrb_int i;
 
   if (t == NULL) return;
   seg = t->rootseg;
   while (seg) {
     for (i=0; i<seg->size; i++) {
-      /* no value in last segment after last_len */
+      /* no value in last h_segment after last_len */
       if (!seg->next && i >= t->last_len) {
         return;
       }
@@ -529,7 +503,7 @@ mrb_hash_foreach(mrb_state *mrb, struct RHash *hash, mrb_hash_foreach_func *func
 static htable*
 ht_copy(mrb_state *mrb, htable *t)
 {
-  segment *seg;
+  h_segment *seg;
   htable *t2;
   mrb_int i;
 
@@ -557,12 +531,12 @@ ht_copy(mrb_state *mrb, htable *t)
 static void
 ht_free(mrb_state *mrb, htable *t)
 {
-  segment *seg;
+  h_segment *seg;
 
   if (!t) return;
   seg = t->rootseg;
   while (seg) {
-    segment *p = seg;
+    h_segment *p = seg;
     seg = seg->next;
     mrb_free(mrb, p);
   }
@@ -1009,7 +983,7 @@ mrb_hash_delete(mrb_state *mrb, mrb_value self)
 static void
 ht_shift(mrb_state *mrb, htable *t, mrb_value *kp, mrb_value *vp)
 {
-  segment *seg = t->rootseg;
+  h_segment *seg = t->rootseg;
   mrb_int i;
 
   while (seg) {
diff --git a/src/variable.c b/src/variable.c
index 030aa7b0..197a83a3 100644
--- a/src/variable.c
+++ b/src/variable.c
@@ -11,23 +11,6 @@
 #include <mruby/string.h>
 #include <mruby/variable.h>
 
-#ifndef MRB_IV_SEGMENT_SIZE
-#define MRB_IV_SEGMENT_SIZE 4
-#endif
-
-typedef struct segment {
-  mrb_sym key[MRB_IV_SEGMENT_SIZE];
-  mrb_value val[MRB_IV_SEGMENT_SIZE];
-  struct segment *next;
-} segment;
-
-/* Instance variable table structure */
-typedef struct iv_tbl {
-  segment *rootseg;
-  size_t size;
-  size_t last_len;
-} iv_tbl;
-
 /* Creates the instance variable table. */
 static iv_tbl*
 iv_new(mrb_state *mrb)
@@ -46,9 +29,9 @@ iv_new(mrb_state *mrb)
 static void
 iv_put(mrb_state *mrb, iv_tbl *t, mrb_sym sym, mrb_value val)
 {
-  segment *seg;
-  segment *prev = NULL;
-  segment *matched_seg = NULL;
+  iv_segment *seg;
+  iv_segment *prev = NULL;
+  iv_segment *matched_seg = NULL;
   size_t matched_idx = 0;
   size_t i;
 
@@ -86,7 +69,7 @@ iv_put(mrb_state *mrb, iv_tbl *t, mrb_sym sym, mrb_value val)
     return;
   }
 
-  seg = (segment*)mrb_malloc(mrb, sizeof(segment));
+  seg = (iv_segment*)mrb_malloc(mrb, sizeof(iv_segment));
   seg->next = NULL;
   seg->key[0] = sym;
   seg->val[0] = val;
@@ -104,7 +87,7 @@ iv_put(mrb_state *mrb, iv_tbl *t, mrb_sym sym, mrb_value val)
 static mrb_bool
 iv_get(mrb_state *mrb, iv_tbl *t, mrb_sym sym, mrb_value *vp)
 {
-  segment *seg;
+  iv_segment *seg;
   size_t i;
 
   if (t == NULL) return FALSE;
@@ -130,7 +113,7 @@ iv_get(mrb_state *mrb, iv_tbl *t, mrb_sym sym, mrb_value *vp)
 static mrb_bool
 iv_del(mrb_state *mrb, iv_tbl *t, mrb_sym sym, mrb_value *vp)
 {
-  segment *seg;
+  iv_segment *seg;
   size_t i;
 
   if (t == NULL) return FALSE;
@@ -158,7 +141,7 @@ iv_del(mrb_state *mrb, iv_tbl *t, mrb_sym sym, mrb_value *vp)
 static void
 iv_foreach(mrb_state *mrb, iv_tbl *t, mrb_iv_foreach_func *func, void *p)
 {
-  segment *seg;
+  iv_segment *seg;
   size_t i;
 
   if (t == NULL) return;
@@ -186,7 +169,7 @@ iv_foreach(mrb_state *mrb, iv_tbl *t, mrb_iv_foreach_func *func, void *p)
 static size_t
 iv_size(mrb_state *mrb, iv_tbl *t)
 {
-  segment *seg;
+  iv_segment *seg;
   size_t size = 0;
 
   if (t == NULL) return 0;
@@ -208,7 +191,7 @@ iv_size(mrb_state *mrb, iv_tbl *t)
 static iv_tbl*
 iv_copy(mrb_state *mrb, iv_tbl *t)
 {
-  segment *seg;
+  iv_segment *seg;
   iv_tbl *t2;
 
   size_t i;
@@ -235,11 +218,11 @@ iv_copy(mrb_state *mrb, iv_tbl *t)
 static void
 iv_free(mrb_state *mrb, iv_tbl *t)
 {
-  segment *seg;
+  iv_segment *seg;
 
   seg = t->rootseg;
   while (seg) {
-    segment *p = seg;
+    iv_segment *p = seg;
     seg = seg->next;
     mrb_free(mrb, p);
   }
```

Reply:
I don't think exposing internal data structure is a wise idea. How about providing a function to get the size of instance variable table e.g. `mrb_obj_iv_memsize()` in `variable.c`?